runtime/src/main/scala/akka/grpc/GrpcProtocol.scala

```diff
@@ -0,0 +1,117 @@
+package akka.grpc
+
+import akka.NotUsed
+import akka.grpc.GrpcProtocol.{ GrpcProtocolMarshaller, GrpcProtocolUnmarshaller }
+import akka.http.javadsl.{ model => jmodel }
+import akka.http.scaladsl.model.{ ContentType, HttpHeader }
+import akka.http.scaladsl.model.HttpEntity.ChunkStreamPart
+import akka.stream.scaladsl.Flow
+import akka.util.ByteString
+
+import scala.util.Try
+
+/**
+ * A variant of the gRPC protocol - e.g. gRPC and gRPC-Web
+ */
+trait GrpcVariant {
+
+  /** The canonical media type to use for this protocol variant */
+  val contentType: ContentType
+
+  /** The set of media types that can identify this protocol variant (e.g. including an implicit +proto) */
+  val mediaTypes: Set[jmodel.MediaType]
+
+  /**
+   * Constructs a marshaller for writing gRPC protocol frames for this variant
+   * @param codec the compression codec to encode data frame bodies with.
+   */
+  def newMarshaller(codec: Codec): GrpcProtocolMarshaller
+
+  /**
+   * Constructs an unmarshaller for reading gRPC protocol frames for this variant.
+   * @param codec the compression codec to decode data frame bodies with.
+   */
+  def newUnmarshaller(codec: Codec): GrpcProtocolUnmarshaller
+}
+
+/**
+ * Core definitions for gRPC protocols.
+ */
+object GrpcProtocol {
+
+  private val grpcVariants: Seq[GrpcVariant] = Seq(Grpc, GrpcWeb, GrpcWebText)
+
+  /** Field marker to signal the start of an uncompressed frame */
+  val notCompressed: ByteString = ByteString(0)
+
+  /** Field marker to signal the start of an frame compressed according to the Message-Encoding from the header */
+  val compressed: ByteString = ByteString(1)
+
+  /** A frame in a logical gRPC protocol stream */
+  sealed trait Frame
+
+  /** A data (or message) frame in a gRPC protocol stream. */
+  case class DataFrame(data: ByteString) extends Frame
+
+  /** A trailer (status headers) frame in a gRPC protocol stream */
+  case class TrailerFrame(trailers: List[HttpHeader]) extends Frame
+
+  /**
+   * Implements the encoding of a stream of gRPC Frames into a physical/transport layer.
+   *
+   * This maps the logical gRPC frames into a stream of chunks that can be handled by the HTTP/2 or HTTP/1.1 transport layer.
+   */
+  case class GrpcProtocolMarshaller(
```

Comment:
I thought this would be contentious (and didn't 100% like Marshaller myself, since that's quite overloaded in the various gRPC frameworks already).

This type is more than just the frame encoder though (that's 2 of the attributes on the class).
The others are the negotiated content type, and the negotiated message encoding.

Reply:
Maybe `GrpcProtocolReader` and `GrpcProtocolWriter` (reader would be used to read requests in the server handler, and responses in the client etc.)

Reply:
I think it would be fine to have the content type and message encoding as properties of the encoding, but I'm OK with reader/writer as well.

Reply:
Have renamed to reader/writer.